src/Config/Converters/DataSourceConverterFactory.cs

```diff
@@ -0,0 +1,148 @@
+                    {
+                        string propertyName = reader.GetString() ?? string.Empty;
+                        reader.Read();
+                        switch (propertyName)
+                        {
+                            case "database-type":
+                                dataSource = dataSource with { DatabaseType = EnumExtensions.Deserialize<DatabaseType>(reader.DeserializeString(_replaceEnvVar)!) };
+                                break;
+                            case "connection-string":
+                                dataSource = dataSource with { ConnectionString = reader.DeserializeString(replaceEnvVar: _replaceEnvVar)! };
+                                break;
+                            case "options":
+                                if (reader.TokenType == JsonTokenType.Null)
+                                {
+                                    dataSource = dataSource with { Options = null };
+                                    break;
+                                }
+                                else
+                                {
+                                    Dictionary<string, JsonElement> optionsDict = new();
+                                    while (reader.Read() && reader.TokenType is not JsonTokenType.EndObject)
+                                    {
+                                        string key = reader.GetString()!;
+                                        reader.Read();
+                                        JsonElement element;
+                                        if (reader.TokenType is JsonTokenType.String)
+                                        {
+                                            string stringValue = reader.DeserializeString(replaceEnvVar: _replaceEnvVar)!;
+
+                                            if (bool.TryParse(stringValue, out bool boolValue))
+                                            {
+                                                // MsSqlOptions.SetSessionContext will contain a boolean value.
+                                                using (JsonDocument doc = JsonDocument.Parse(boolValue.ToString().ToLower()))
+                                                {
+                                                    element = doc.RootElement.Clone();
+                                                }
+                                            }
+                                            else
+                                            {
+                                                // CosmosDbNoSQLDataSourceOptions will contain string values.
+                                                using (JsonDocument doc = JsonDocument.Parse($"\"{stringValue}\""))
+                                                {
+                                                    element = doc.RootElement.Clone();
+                                                }
+                                            }
```

Comment:
Wouldn't you already know the db type at this point of json parsing? I suppose only if the user places 'database-type` before the `options` property.

So if user tries to provide a bool for `CosmosDbNoSQLDataSourceOptions`, we could error out with appropriate message?

Reply:
The role of the serializer is simply to correctly deserialize the config containing env values. All the checks on the reuqired options for a specific db is already done separately. we don't require to have it at multiple places.